tiledb/type/range/range.h

```diff
@@ -257,6 +259,102 @@ class Range {
   uint64_t partition_depth_;
 };
 
+/**
+ * Performs correctness checks for a valid range. If any validity checks fail
+ * an error status is returned.
+ *
+ * @param range The range to check.
+ * @return Status of the checks.
+ */
+template <
+    typename T,
+    typename std::enable_if<std::is_arithmetic<T>::value>::type* = nullptr>
+Status check_range_is_valid(const Range& range) {
+  // Check has data.
+  if (range.empty())
+    return Status_Error("Range is empty");
+  auto r = (const T*)range.data();
+  // Check for NaN
+  if constexpr (std::is_floating_point_v<T>) {
+    if (std::isnan(r[0]) || std::isnan(r[1]))
+      return Status_Error("Range contains NaN");
+  }
+  // Check range bounds
+  if (r[0] > r[1]) {
+    std::stringstream ss;
+    ss << "Lower range bound " << r[0]
+       << " cannot be larger than the higher bound " << r[1];
+    return Status_Error(ss.str());
+  }
+  // Return okay
+  return Status::Ok();
+};
+
+/**
+ * Replaces a range with its intersection with the superset.
+ *
+ * This method will return an ok status if the Range is not mutated and an
+ * error status if the intersection changes the bounds of the range. Both the
+ * superset and the range must be valid ranges of the appropriate types as
+ * checked by ``check_typed_range_is_valid``.
+ *
+ * @param superset The range to intersect with. Must be a valid range
+ * with data of type T.
+ * @param range The range to replace with the intersection. Must be a valid
+ * range with data of type T.
+ * @return Status that returns an error if range was mutated.
+ */
+template <
+    typename T,
+    typename std::enable_if<std::is_arithmetic<T>::value>::type* = nullptr>
+Status intersect_range(const Range& superset, Range& range) {
```

Comment:
This function is quite confusing. It mutates the range, yet returns an error? I think we should clarify this much better or split it into `intersects` that returns `bool` and `crop` that returns `Range`. Also I am not sure about variable name `superset`, since in the case of mutation/error `superset` is not a superset after all.

Reply:
Updated.